saleor/order/utils.py

```diff
@@ -52,20 +52,28 @@ def add_items_to_delivery_group(delivery_group, partition, discounts=None):
     for item_line in partition:
         product_variant = item_line.variant
         price = item_line.get_price_per_item(discounts)
-        quantity = item_line.get_quantity()
-        stock = product_variant.select_stockrecord(quantity)
-        delivery_group.items.create(
-            product=product_variant.product,
-            quantity=quantity,
-            unit_price_net=price.net,
-            product_name=product_variant.display_product(),
-            product_sku=product_variant.sku,
-            unit_price_gross=price.gross,
-            stock=stock,
-            stock_location=stock.location.name if stock else None)
-        if stock:
-            # allocate quantity to avoid overselling
-            Stock.objects.allocate_stock(stock, quantity)
+        total_quantity = item_line.get_quantity()
+
+        while total_quantity > 0:
```

Comment:
I'm afraid that we might end up with an infinite loop here. We allow to perform various operations on orders that change the quantity, if somehow the stock quantities go out of sync, this would break. Maybe something like `if not stock: break` would help?

Reply:
You are right. Although we are checking available quantity earlier, it would be much safier with this condition.

Reply:
I missed one thing - it's already solved in few lines below. When stock is not found, then item is added with all left quantity and no stock. Then problem with missing products can be manually solved by staff users and in dashboard still appears quantity chosen by customer.

Reply:
I think we should fail loudly, rolling back database changes and displaying an error rather than leave the system in an inconsistent state and have staff figure out how to fix things. In a large store the person handling orders may not have access or even ability to fix stock.

Reply:
For now, I solved it by throwing an ```InsufficientStock``` exception here if there is no stock. It is handled in ```handle_order_placement``` method, which redirects user back to cart. Then it is displayed standard message: ```Sorry. We don't have that many items in stock. Quantity was set to maximum available for now.```.